river/LayoutManager.zig

```diff
@@ -0,0 +1,300 @@
+// This file is part of river, a dynamic tiling wayland compositor.
+//
+// Copyright 2020 - 2021 The River Developers
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+const Self = @This();
+
+const std = @import("std");
+const wlr = @import("wlroots");
+const wayland = @import("wayland");
+const wl = wayland.server.wl;
+const zriver = wayland.server.zriver;
+
+const log = @import("log.zig");
+const util = @import("util.zig");
+
+const Server = @import("Server.zig");
+const Output = @import("Output.zig");
+
+const Error = error{BindFailed};
+
+pub const LayoutDemand = struct {
+    serial: u32 = 0,
+
+    // TODO we need a list of ViewDimensions which has a static size,
+    //      so nodes can be added at arbitrary indices
+
+    /// Timeout timer
+    timer: ?*wl.EventSource = null,
+
+    /// Arbitrary timeout
+    const timeout: u32 = 1000;
+
+    const ViewDimensions = struct {
+        x: i32, y: i32, w: u32, h: u32, index: u32
+    };
+
+    pub fn init(self: *LayoutDemand) void {
+        const output = @fieldParentPtr(Output, "layout_demand", self);
```

Comment:
you can't use `@fieldParentPtr()` here as `Output` stores a pointer to a `LayoutDemand` not the struct itself